public/js/reducers/sources.js

```diff
@@ -54,19 +54,19 @@ function update(state = State(), action: Action) : Record<SourcesState> {
 
     case "SELECT_SOURCE":
       return state.merge({
-        selectedSource: action.source,
         pendingSelectedSourceURL: null,
-        tabs: updateTabList(state, fromJS(action.source), action.options)
-      });
+        tabs: updateTabList(state, fromJS(action.source), action.tabIndex)
+      }).set("selectedLocation", { sourceId: action.source.id,
+                                   line: action.line });
 
     case "SELECT_SOURCE_URL":
       return state.merge({ pendingSelectedSourceURL: action.url });
 
     case "CLOSE_TAB":
-      return state.merge({
-        selectedSource: getNewSelectedSource(state, action.id),
-        tabs: removeSourceFromTabList(state, action.id)
-      });
+      return state.merge({ tabs: removeSourceFromTabList(state, action.id) })
+        .set("selectedLocation", {
+          sourceId: getNewSelectedSource(state, action.id)
+        });
```

Comment:
out of curiosity, why not keep this as one merge? I'm probably missing something :)